File: scratch/sync_agent.py
import os
import time
import requests
import json
import subprocess
import tkinter as tk
from tkinter import messagebox, filedialog
import threading

# Config
SERVER_URL = "http://140.245.76.33"  # Oracle VM IP
SYNC_DIR = os.path.join(os.path.expanduser("~"), "Desktop", "KFC_MAN_Sync")
os.makedirs(SYNC_DIR, exist_ok=True)

class SyncAgentApp:
    def __init__(self, root):
        self.root = root
        self.root.title("KFC MAN.HWP - 실시간 동기화 에이전트")
        self.root.geometry("480x360")
        self.root.configure(bg="#0f172a")
        
        self.watched_files = {}  # {filename: (doc_id, last_mtime)}
        self.running = True
        
        # UI Setup
        self.title_label = tk.Label(root, text="🍀 KFC MAN.HWP 폴더 동기화 에이전트", font=("Pretendard", 14, "bold"), fg="#38bdf8", bg="#0f172a")
        self.title_label.pack(pady=15)
        
        self.dir_label = tk.Label(root, text=f"📂 감시 폴더: {SYNC_DIR}", font=("Pretendard", 10), fg="#94a3b8", bg="#0f172a")
        self.dir_label.pack(pady=5)
        
        self.log_box = tk.Text(root, height=12, width=55, bg="#1e293b", fg="#e2e8f0", font=("Consolas", 9), bd=0)
        self.log_box.pack(pady=10)
        
        self.btn_open = tk.Button(root, text="공유 폴더 열기", command=self.open_folder, bg="#3b82f6", fg="white", font=("Pretendard", 9, "bold"), padx=10, pady=5)
        self.btn_open.pack(side=tk.LEFT, padx=30, pady=10)
        
        self.btn_sync = tk.Button(root, text="즉시 클라우드 동기화", command=self.force_sync, bg="#10b981", fg="white", font=("Pretendard", 9, "bold"), padx=10, pady=5)
        self.btn_sync.pack(side=tk.RIGHT, padx=30, pady=10)
        
        # Start Threads
        self.log("동기화 시스템이 작동을 시작했습니다.")
        self.log("kfcman.link 와 바탕화면 KFC_MAN_Sync 폴더를 감시합니다.")
        
        threading.Thread(target=self.sync_loop, daemon=True).start()
        
    def log(self, msg):
        self.log_box.insert(tk.END, f"[{time.strftime('%H:%M:%S')}] {msg}\n")
        self.log_box.see(tk.END)
        
    def open_folder(self):
        os.startfile(SYNC_DIR)
        
    def force_sync(self):
        self.log("🔄 서버로부터 새로운 문서 목록을 수신하여 동기화하는 중...")
        try:
            res = requests.get(f"{SERVER_URL}/api/docs")
            if res.status_code == 200:
                docs = res.json().get("docs", [])
                hwp_docs = [d for d in docs if d.get("hasHwpData")]
                self.log(f"발견된 클라우드 한글 문서: {len(hwp_docs)}개")
                
                for doc in hwp_docs:
                    doc_id = doc["id"]
                    filename = doc["hwpName"] or f"{doc['title']}.hwp"
                    file_path = os.path.join(SYNC_DIR, filename)
                    
                    # Download file if not exists
                    if not os.path.exists(file_path):
                        self.log(f"📥 다운로드 중: {filename}")
                        down_res = requests.get(f"{SERVER_URL}/api/docs/{doc_id}/download")
                        if down_res.status_code == 200:
                            with open(file_path, "wb") as f:
                                f.write(down_res.content)
                            mtime = os.path.getmtime(file_path)
                            self.watched_files[filename] = (doc_id, mtime)
                            self.log(f"✅ 다운로드 완료: {filename}")
            else:
                self.log("❌ 문서 목록 수신 실패")
        except Exception as e:
            self.log(f"❌ 동기화 에러: {str(e)}")
            
    def sync_loop(self):
        # Initial force sync
        self.force_sync()
        
        while self.running:
            time.sleep(2)
            try:
                # 1. Check local changes to upload
                for filename in os.listdir(SYNC_DIR):
                    if not (filename.endswith(".hwp") or filename.endswith(".hwpx")):
                        continue
                    
                    file_path = os.path.join(SYNC_DIR, filename)
                    mtime = os.path.getmtime(file_path)
                    
                    # If this is a new file or modified file
                    if filename not in self.watched_files:
                        # Skip initial adding if it was already handled by download
                        continue
                        
                    doc_id, last_mtime = self.watched_files[filename]
                    if mtime > last_mtime:
                        self.log(f"⚡ 변경사항 감지! 업로드 중: {filename}")
                        with open(file_path, "rb") as f:
                            import base64
                            binary_data = f.read()
                            base64_data = "data:application/x-hwp;base64," + base64.b64encode(binary_data).decode('utf-8')
                        
                        payload = {
                            "title": filename,
                            "hwpData": base64_data,
                            "hwpName": filename
                        }
                        res = requests.put(f"{SERVER_URL}/api/docs/{doc_id}", json=payload)
                        if res.status_code == 200:
                            self.watched_files[filename] = (doc_id, mtime)
                            self.log(f"✅ 클라우드 동기화 및 덮어쓰기 완료!")
                        else:
                            self.log(f"❌ 동기화 실패: {res.status_code}")
                            
            except Exception as e:
                pass

if __name__ == "__main__":
    root = tk.Tk()
    app = SyncAgentApp(root)
    root.mainloop()

File: scratch/sync_agent.js
const fs = require('fs');
const path = require('path');
const http = require('http');
const { exec } = require('child_process');

const SERVER_URL = "http://140.245.76.33";  // Oracle VM IP
const SYNC_DIR = path.join(require('os').homedir(), 'Desktop', 'KFC_MAN_Sync');

if (!fs.existsSync(SYNC_DIR)) {
  fs.mkdirSync(SYNC_DIR, { recursive: true });
}

let watchedFiles = {}; // {filename: {doc_id, last_mtime}}

console.log("==================================================");
console.log("🍀 KFC MAN.HWP 실시간 로컬 폴더 동기화 에이전트");
console.log(`📂 감시 대상 바탕화면 폴더: ${SYNC_DIR}`);
console.log("==================================================");

function log(msg) {
  const time = new Date().toLocaleTimeString();
  console.log(`[${time}] ${msg}`);
}

// REST helper to fetch json
function fetchJson(url) {
  return new Promise((resolve, reject) => {
    http.get(url, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          resolve(JSON.parse(data));
        } catch (e) {
          reject(e);
        }
      });
    }).on('error', reject);
  });
}

// Download file helper
function downloadFile(docId, filename) {
  const fileUrl = `${SERVER_URL}/api/docs/${docId}/download`;
  const fileWriter = fs.createWriteStream(path.join(SYNC_DIR, filename));
  
  return new Promise((resolve, reject) => {
    http.get(fileUrl, (res) => {
      res.pipe(fileWriter);
      fileWriter.on('finish', () => {
        fileWriter.close();
        const mtime = fs.statSync(path.join(SYNC_DIR, filename)).mtimeMs;
        resolve(mtime);
      });
    }).on('error', reject);
  });
}

// Force sync docs from server
async function forceSync() {
  log("🔄 서버로부터 한글 파일 목록 동기화 중...");
  try {
    const data = await fetchJson(`${SERVER_URL}/api/docs`);
    const docs = data.docs || [];
    const hwpDocs = docs.filter(d => d.hasHwpData);
    
    log(`발견된 클라우드 문서: ${hwpDocs.length}개`);
    
    for (const doc of hwpDocs) {
      const docId = doc.id;
      const filename = doc.hwpName || `${doc.title}.hwp`;
      const filePath = path.join(SYNC_DIR, filename);
      
      if (!fs.existsSync(filePath)) {
        log(`📥 새 문서 다운로드 중: ${filename}`);
        const mtime = await downloadFile(docId, filename);
        watchedFiles[filename] = { doc_id: docId, last_mtime: mtime };
        log(`✅ 다운로드 완료: ${filename}`);
      } else if (!watchedFiles[filename]) {
        // If file exists locally but not in memory list, record its mtime
        const mtime = fs.statSync(filePath).mtimeMs;
        watchedFiles[filename] = { doc_id: docId, last_mtime: mtime };
      }
    }
  } catch (err) {
    log(`❌ 동기화 중 오류 발생: ${err.message}`);
  }
}

// Put request upload helper
function uploadFile(docId, filename, filePath) {
  return new Promise((resolve, reject) => {
    const fileBuffer = fs.readFileSync(filePath);
    const base64Data = "data:application/x-hwp;base64," + fileBuffer.toString('base64');
    
    const payload = JSON.stringify({
      title: filename,
      hwpData: base64Data,
      hwpName: filename
    });
    
    const url = new URL(`${SERVER_URL}/api/docs/${docId}`);
    const req = http.request({
      hostname: url.hostname,
      port: url.port || 80,
      path: url.pathname,
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(payload)
      }
    }, (res) => {
      let responseBody = '';
      res.on('data', chunk => responseBody += chunk);
      res.on('end', () => {
        if (res.statusCode === 200) {
          resolve(true);
        } else {
          reject(new Error(`Status ${res.statusCode}: ${responseBody}`));
        }
      });
    });
    
    req.on('error', reject);
    req.write(payload);
    req.end();
  });
}

// Watch loop
async function watchLoop() {
  await forceSync();
  
  setInterval(async () => {
    try {
      const files = fs.readdirSync(SYNC_DIR);
      for (const filename of files) {
        if (! (filename.endsWith('.hwp') || filename.endsWith('.hwpx')) ) continue;
        
        const filePath = path.join(SYNC_DIR, filename);
        const stats = fs.statSync(filePath);
        const mtime = stats.mtimeMs;
        
        if (!watchedFiles[filename]) continue;
        
        const { doc_id, last_mtime } = watchedFiles[filename];
        if (mtime > last_mtime) {
          log(`⚡ 바탕화면 한글 파일 변경 감지! 업로드 중: ${filename}`);
          try {
            await uploadFile(doc_id, filename, filePath);
            watchedFiles[filename].last_mtime = mtime;
            log(`✅ 클라우드 업로드 동기화 완료!`);
          } catch (uploadErr) {
            log(`❌ 업로드 실패: ${uploadErr.message}`);
          }
        }
      }
    } catch (e) {
      // Quiet errors
    }
  }, 2000);
}

// Open folder natively
exec(`explorer "${SYNC_DIR}"`);

// Run
watchLoop();
